### PR TITLE
CAF-2890: Update workers to pause when they are unhealthy

### PR DESCRIPTION
**Pull request by glucy**
https://jira.autonomy.com/browse/CAF-2890

A failing health checks will now stop a worker consuming messages from the workers input queue.
Once all health checks pass successfully the worker will resume consuming messages from its input queue.

### CODE REVIEW

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthCheck.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        this.healthCheck = healthCheck;
+    }
+
+    @Override
+    /**
+     * This method executes the available health checks and can disconnect or reconnect a worker to
+     * it's input queue depending on the workers overall health.
+     * 
+     * If a health check fails the health check is added to the set of unhealthy checks and the
+     * worker is disconnecting from it's input queue.
```
Comment (arogan92) [inline worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthCheck.java]: TYPO: disconnected

**Review by mulhollandc — APPROVED** [review]

**Review by arogan92 — COMMENTED** [review]
Just a typo
Code looks good

File: worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthCheck.java (hunk 1 above)
Comment (glucy) [reply worker-core/src/main/java/com/hpe/caf/worker/core/GatedHealthCheck.java]: Already fixed by Connor 👍

**Comment by glucy** [comment]
@dermot-hardy I've added the additional functionality so works experiencing a transient failure will check health checks before taking on more work.

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerApplication.java [hunk 2]
```diff
@@ -119,8 +118,8 @@ public void run(final WorkerConfiguration workerConfiguration, final Environment
         ManagedWorkerQueue workerQueue = queueProvider.getWorkerQueue(config, nThreads);
         MessagePriorityManagerProvider priorityManagerProvider = ModuleLoader.getService(MessagePriorityManagerProvider.class);
         MessagePriorityManager priorityManager = priorityManagerProvider.getMessagePriorityManager(config);
-
-        WorkerCore core = new WorkerCore(codec, wtp, workerQueue, priorityManager, workerFactory, path);
+        TransientHealthCheck transientHealthCheck = new TransientHealthCheck();
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerApplication.java]: Extra whitespace at the end of this line.

File: worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java [hunk 3]
```diff
@@ -405,17 +410,20 @@ public void complete(final String queueMsgId, final String queue, final TaskMess
...
                 LOG.error("Cannot publish data for task {}, rejecting", responseMessage.getTaskId(), e);
-                abandon(queueMsgId);
+                abandon(queueMsgId, e);
             }
         }
 
 
         @Override
-        public void abandon(final String queueMsgId)
+        public void abandon(final String queueMsgId, Exception e)
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/WorkerCore.java]: Parameter should be `final`.

File: worker-core/src/main/java/com/hpe/caf/worker/core/TransientHealthCheck.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                
+                healthResult = HealthResult.RESULT_HEALTHY;
+            } else {
+                healthResult = new HealthResult(HealthStatus.UNHEALTHY);
+                break;
+            }
+        }
+        if(healthResult.equals(HealthResult.RESULT_HEALTHY)) {
+            LOG.debug("Transient Health Check is currently Healthy, therefore clearing the Transient Exception Registry");
+            transientExceptionRegistry.clear();
```
Comment (dermot-hardy) [inline worker-core/src/main/java/com/hpe/caf/worker/core/TransientHealthCheck.java]: Thread safety issue here as discussed.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

**Review by dermot-hardy — APPROVED** [review]

**Pull request merged**